lib/Dialect/FIRRTL/Transforms/ExtractInstances.cpp

```diff
@@ -0,0 +1,964 @@
+    auto name = getModuleNamespace(module).newName("extraction_sym");
+    attr = StringAttr::get(op->getContext(), name);
+    op->setAttr("inner_sym", attr);
+    return attr;
+  }
+
+  /// Obtain an inner reference to an operation, possibly adding an `inner_sym`
+  /// to that operation.
+  InnerRefAttr getInnerRefTo(Operation *op) {
+    return InnerRefAttr::get(
+        SymbolTable::getSymbolName(op->getParentOfType<FModuleOp>()),
+        getOrAddInnerSym(op));
+  }
+
+private:
+  bool anythingChanged;
+  bool anyFailures;
+
+  InstanceGraph *instanceGraph = nullptr;
+
+  /// The modules in the design that are annotated with one or more annotations
+  /// relevant for instance extraction.
+  DenseMap<Operation *, SmallVector<Annotation, 1>> annotatedModules;
+
+  /// All modules that are marked as DUT themselves. Realistically this is only
+  /// ever one module in the design.
+  DenseSet<Operation *> dutRootModules;
+  /// All modules that are marked as DUT themselves, or that have a DUT parent
+  /// module.
+  DenseSet<Operation *> dutModules;
+  /// All DUT module names.
+  DenseSet<Attribute> dutModuleNames;
+
+  /// A worklist of instances that need to be moved.
+  SmallVector<std::pair<InstanceOp, ExtractionInfo>> extractionWorklist;
+
+  /// The path along which instances have been extracted. This essentially
+  /// documents the original location of the instance in reverse. Every push
+  /// upwards in the hierarchy adds another entry to this path documenting along
+  /// which instantiation path each instance was extracted.
+  DenseMap<Operation *, SmallVector<InnerRefAttr>> extractionPaths;
+
+  /// A map of the original parent modules of instances before they were
+  /// extracted. This is used in a corner case during trace file emission.
+  DenseMap<Operation *, StringAttr> originalInstanceParents;
+
+  /// All extracted instances in their position after moving upwards in the
+  /// hierarchy, but before being grouped into an optional submodule.
+  SmallVector<std::pair<InstanceOp, ExtractionInfo>> extractedInstances;
+
+  // The uniquified wiring prefix for each instance.
+  DenseMap<Operation *, SmallString<16>> instPrefices;
+
+  /// The current circuit namespace valid within the call to `runOnOperation`.
+  CircuitNamespace circuitNamespace;
+  /// Cached module namespaces.
+  DenseMap<Operation *, ModuleNamespace> moduleNamespaces;
+};
+} // end anonymous namespace
+
+/// Emit the annotated source code for black boxes in a circuit.
+void ExtractInstancesPass::runOnOperation() {
+  CircuitOp circuitOp = getOperation();
+  anythingChanged = false;
+  anyFailures = false;
+  annotatedModules.clear();
+  dutRootModules.clear();
+  dutModules.clear();
+  extractionWorklist.clear();
+  extractionPaths.clear();
+  originalInstanceParents.clear();
+  extractedInstances.clear();
+  instPrefices.clear();
+  moduleNamespaces.clear();
+  circuitNamespace.clear();
+  circuitNamespace.add(circuitOp);
```

Comment:
The clearing of all this mutable state is hygenic, but likely an anti-pattern. A follow-up cleanup would be to rework all the mutable state into a separate class/struct that only lives for the lifetime of `runOnOperation`.

Reply:
Yeah I totally agree. There was some rather esoteric performance argument for this, where multiple runs of the pass would reuse the allocated memory of a prior pass, so you'd only pay for the gradual growing of underlying storage once. Totally doesn't apply here though...